### ISSUE TITLE
NodeMaterial : can't extend NodeMaterials classes

### ISSUE THREAD

**Issue by wmcmurray** [mrdoob/three.js#23707]
**Describe the bug**

Extending any of the NodeMaterial based material classes creates a broken and unusable material since r138. It gives me this error : `Uncaught TypeError: Cannot set properties of undefined (setting 'value')`.

**Fix the bug**

This problem occurs because the [constructor.name is used to determine the value of type](https://github.com/mrdoob/three.js/blob/98985923afe4fe4030fce256b0cc0d02dc68696f/examples/jsm/renderers/nodes/materials/NodeMaterial.js#L10). This value will equal the name of the top-most class in the inheritance chain.

I believe setting it manually would be better (ex: `PointsNodeMaterial.prototype.type = 'PointsNodeMaterial';`).
Or using a boolean (ex: `PointsNodeMaterial.prototype.isPointsNodeMaterial = true;`) (like many other things in threejs) !

***Live example***

* [codepen-latest-release](https://codepen.io/wmcmurray/pen/oNpvamN?editors=1010) (line 50 👀 )

**Expected behavior**

Extending any ThreeJs material class should work out of the box.

**Platform:**

 - Device: all
 - OS: all
 - Browser: all
 - Three.js version: r138

**Comment by sunag** [comment]
[I posted a PR to fix it.](https://github.com/mrdoob/three.js/pull/23724) I think it would be better not to use types other than what the class represents because that breaks serialization.

**Comment by wmcmurray** [comment]
This bug is back 😬 because [the fix provided](https://github.com/mrdoob/three.js/pull/23724) has been [removed in this commit](https://github.com/mrdoob/three.js/pull/24453/commits/16d6c702edac79b20afb8ea2b65a8e1701e0396c#diff-c1dd4865c519271fcfbc0edc7f537658d0e1aa517cef9aff618fcd7294f901ddL70-L79) (see `WebGLNodeBuilder.js`).
And that fix was just a workaround anyway, it did not fix the problem [directly where it is](https://github.com/mrdoob/three.js/blame/9d6c37f6acfd62672fd1343b9d7a028142c6af8d/examples/jsm/nodes/materials/NodeMaterial.js#L19).

Let me illustrate what is happening (just to be clear):

```js
class NodeMaterial {
  constructor() {
    this.type = this.constructor.name;
  }
}

class MySpecialMaterial extends NodeMaterial {
  // ...
}

const resultA = new NodeMaterial();
const resultB = new MySpecialMaterial();

console.log(resultA.type); // will output: "NodeMaterial", ✅ good
console.log(resultB.type); // will output: "MySpecialMaterial", ❌ bad !
```

(PS: `constructor.name` is also used [here](https://github.com/mrdoob/three.js/blame/9d6c37f6acfd62672fd1343b9d7a028142c6af8d/examples/jsm/nodes/core/Node.js#L23-L27), but this one doesn't seem to have any side effects (in my project at least))

**Comment by sunag** [comment]
Ohh.. I will check this.

**Comment by sunag** [comment]
> And that fix was just a workaround anyway, it did not fix the problem [directly where it is](https://github.com/mrdoob/three.js/blame/9d6c37f6acfd62672fd1343b9d7a028142c6af8d/examples/jsm/nodes/materials/NodeMaterial.js#L19).

The behavior is correct here, no extended `Node` or `Material` should use a `type` other than the name of the class that builds it, this breaks any serious serialization system.

**Comment by wmcmurray** [comment]
> no extended `Node` or `Material` should use a `type` other than the name of the class that builds it

Agreed, this is exactly my point ! 😅  and why I think [fixing it like this](https://github.com/mrdoob/three.js/pull/24723/commits/3a1f022f78fb3b0cde1da2f479c3b85558690edb) is not the proper solution.

Let me explain:
During the class construction phase, if you set the type with `this.type = this.constructor.name;`, **it will be wrong** after instanciation (in the context of an extended class, like my `MySpecialMaterial` example above), up until the `WebGLNodeBuilder._parseShaderLib()` method get called during the [Material.onBuild()](https://github.com/mrdoob/three.js/blob/9cfb9790cafcd9728800d4392b4b9d6fe16a6746/examples/jsm/renderers/webgl/nodes/WebGLNodes.js#L13) phase. This creates an invisible coupling that can cause a lot of unpredictable problems, like what just happened: someone remove that piece of code that fixes the type value, and things unexpectedly break elsewhere. Other problems could arise if an other piece of code checks the value of `material.type` (because it still have an invalid value, it's only "fixed" inside of `_parseShaderLib()`).

It would be much more solid and predictable to define the type this way:

```js
class LineBasicNodeMaterial extends NodeMaterial {
	constructor() {
		this.type = 'LineBasicNodeMaterial';
	}
}

class PointsNodeMaterial extends NodeMaterial {
	constructor() {
		this.type = 'PointsNodeMaterial';
	}
}

class MeshPhysicalNodeMaterial extends NodeMaterial {
	constructor() {
		this.type = 'MeshPhysicalNodeMaterial';
	}
}

// ...etc
```

There is no invisible coupling there, no unpredictable side effects, etc.

**Comment by sunag** [comment]
This is generated automatically and the way it is serialized, I wouldn't want to leave a "short blanket" here and solve one thing and leave another aside, the current approach meets both demands. `type` is always the class that built it so you can use `isNodeMaterial` to check if it's an extended class.

You shouldn't have problems with `Node` because its architecture is already updated, did you imagine how many extra strings we will have to add for each node following this model? Besides being something else for the user to worry about that they really shouldn't since we can get the class extended with `.isNode` which is much more efficient and we should also think about doing procedurally.

If you need a provisional workaround still you can added:

```
class MySpecialMaterial extends MeshPhysicalNodeMaterial {
	get type() {
		return 'MeshPhysicalNodeMaterial';
	}
}
```

By doing so you will not be able to deserialize this.

**Comment by wmcmurray** [comment]
The `.isSomething` approach should always be used internally then, which reduce the relevance of the `type` property..
is it only used for serialization ? or is there other use for it ?

Because if it's sole purpose is serialization, there is no point in being able to store that the type equals `MySpecialMaterial` inside a serialized format if Three.js can't even unserialize it ! (Three.js does not know where the `MySpecialMaterial` class is, it's part of my application, not part of Three.js, it won't be able to instanciate it).

In other words: If a user want to implement a custom material and also use the serialization mechanism, it **would have to** provide an unserialization mechanism for his custom material, it also make sense to let that user decide what `type` his material should have (because it's his concern, not Three.js's concern). _Separating concerns goes a long way in software architecture._

Besides, just search for `this.type =` inside the `/src` directory of Three.js, you'll notice that pretty much every `type` properties are defined with a manually typed string. It might seem tedious, but when the project has like 85 000 github stars, the priority should be to make sure the codebase is stable for all those users, not saving 10 minutes for one developer, I think :)

**Comment by sunag** [comment]
Sorry, but there seems to be a diversity of opinions here, I wouldn't like to confuse bugs and different types of approach, so if other problems arise in this sense we can review this, I think 85k was achieved because Three.js is always open to improvement...

**Comment by wmcmurray** [comment]
Well then 🤷🏼‍♂️

**Comment by mrdoob** [comment]
> It might seem tedious, but when the project has like 85 000 github stars, the priority should be to make sure the codebase is stable for all those users, not saving 10 minutes for one developer, I think :)

Please don't give use that kind of arguments. We're doing our best 😅

Any chance you can explain the problem in a more clear and concise way?
I'm having a hard time reading your posts...

**Comment by wmcmurray** [comment]
That one slipped, sorry ! 🫶 😬 

I guess everything is fine for now, **[his PR](https://github.com/mrdoob/three.js/pull/24723) probably patched the issue**.

I'm just afraid this issue could come back (a third time) if we are still using `this.constructor.name` to define `this.type`, because it gives an unexpected value when the class is extended.

The solution is yet very simple, we just need to define `this.type` with a manually typed string instead of the dynamic value provided by `this.constructor.name` (like [it's done](https://github.com/mrdoob/three.js/blob/e732f0ca0f19372e1bf0fb12313a7ecaa45b73e3/src/geometries/BoxGeometry.js#L11) [everywhere](https://github.com/mrdoob/three.js/blob/e732f0ca0f19372e1bf0fb12313a7ecaa45b73e3/src/core/Object3D.js#L41) [else](https://github.com/mrdoob/three.js/blob/e732f0ca0f19372e1bf0fb12313a7ecaa45b73e3/src/materials/Material.js#L20) [in the](https://github.com/mrdoob/three.js/blob/e732f0ca0f19372e1bf0fb12313a7ecaa45b73e3/src/materials/MeshPhysicalMaterial.js#L21) [codebase](https://github.com/mrdoob/three.js/blob/e732f0ca0f19372e1bf0fb12313a7ecaa45b73e3/src/materials/PointsMaterial.js#L12)).

**Comment by sunag** [comment]
When I did the [new `WebGLNodeBuilder`](https://github.com/mrdoob/three.js/pull/24453) I must not have synced all the updates, this is one of the "big updates" problems. One of the safest methods of testing Three.js is the examples, and then the unit test that we are yet to implement for the nodes. The reason this has been repeated is the lack of an example that we can test after updates of "high impact" like new `WebGLNodeBuilder`, for that reason I added a[ simple example in this PR](https://github.com/mrdoob/three.js/pull/24723/files#diff-529a005ecfd1bfd1f80ce19ea14c78b41aced39e2a41796557ebbdf19886c467) to avoid repeating this type of problem.
